### PR TITLE
Implement Client with BackOff

### PR DESCRIPTION
**Pull request by bonds0097**
This implements a Client that follows a BackOff policy which allows for stuff like exponential back-off on failures from an http request, allowing you to invisibly deal with timeouts or other errors.

This also re-implements the retry client as a backoff client.

### CODE REVIEW

File: httpretry.go [hunk 1]
```diff
@@ -2,64 +2,29 @@ package httpmock
...
 }
 
 // NewRetryClient returns a wrapped Client that retries request up to retries
 // times.
 func NewRetryClient(client Client, retries int) Client {
-	return &ClientWithRetry{
-		client:  client,
-		retries: retries,
+	return &clientWithRetry{
+		client: clientWithBackOff{
```
Comment (rikonor) [inline httpretry.go]: You've kinda made `clientWithRetry` redundant here. Couldn't `NewRetryClient` just return `clientWithBackOff` now?

Comment (bonds0097) [inline httpretry.go]: Yeah, I guess so.

**Comment by bonds0097** [comment]
What do you mean though? Do you not know what a back-off policy is or what? The way you use the feauture is by creating a backoff client with a backoff policy, otherwise it behaves the same as any other `Client` from an outside perspective.

I don't understand what you think is unclear but if you explain perhaps I can improve docs.

**Comment by rikonor** [comment]
This is the example that you provide:
```go
client := httpmock.NewBackOffClient(
  &http.Client{Timeout: 1 * time.Second},
  &backoff.NewExponentialBackOff(),
)
```
but `backoff.NewExponentialBackOff` is not covered in these docs. So to understand what is happening in practice (even if one is familiar with the concept of exponential backoff) one will have to go to `backoff`'s docs to get a complete understanding.

Additionally, this change would allow a user to define his own `Backoff`, which is what I meant we should be showing an example of.

**Comment by bonds0097** [comment]
Well, I certainly wasn't going to re-implement exponential back-off in my example. Like, you don't have to know anything about the implementation itself, just that it does exponential back-off. 

And I'm also not implementing some other backoff policy in the examples, there is a custom one defined in the library itself that a user can examine and it has relevant docstrings.

**Comment by bonds0097** [comment]
I linked to the backoff repo in the readme under the expectation that anyone who wants to understand how backoffs work in go will go look at that. If they don't want to, that's fine. But I'm not holding anyone's hand here.

**Pull request merged**